src/core/net/dns_client.cpp

```diff
@@ -43,357 +43,777 @@
 }
 
-void Client::HandleUdpReceive(Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+void Client::ProcessResponse(const Message &aMessage)
 {
-    // RFC1035 7.3. Resolver cannot rely that a response will come from the same address
-    // which it sent the corresponding query to.
-    OT_UNUSED_VARIABLE(aMessageInfo);
+    Response  response;
+    QueryType type;
+    otError   responseError;
 
-    otError       error = OT_ERROR_NOT_FOUND;
-    Header        responseHeader;
-    QueryMetadata queryMetadata;
-    AaaaRecord    record;
-    Message *     message = nullptr;
-    uint16_t      offset  = aMessage.GetOffset();
+    response.mMessage = &aMessage;
 
-    SuccessOrExit(aMessage.Read(offset, responseHeader));
-    VerifyOrExit(responseHeader.GetType() == Header::kTypeResponse && responseHeader.GetQuestionCount() == 1 &&
-                 !responseHeader.IsTruncationFlagSet());
-    offset += sizeof(responseHeader);
+    SuccessOrExit(ParseResponse(response, type, responseError));
+    FinalizeQuery(response, type, responseError);
 
-    VerifyOrExit((message = FindQueryById(responseHeader.GetMessageId())) != nullptr);
-    queryMetadata.ReadFrom(*message);
+exit:
+    return;
+}
 
-    VerifyOrExit(responseHeader.GetResponseCode() == Header::kResponseSuccess, error = OT_ERROR_FAILED);
+otError Client::ParseResponse(Response &aResponse, QueryType &aType, otError &aResponseError)
+{
+    otError        error   = OT_ERROR_NONE;
+    const Message &message = *aResponse.mMessage;
+    uint16_t       offset  = message.GetOffset();
+    Header         header;
+    QueryInfo      info;
 
-    // Parse and check the question section.
-    SuccessOrExit(error = CompareQuestions(aMessage, *message, offset));
+    SuccessOrExit(error = message.Read(offset, header));
+    offset += sizeof(Header);
 
-    // Parse and check the answer section.
-    for (uint32_t index = 0; index < responseHeader.GetAnswerCount(); index++)
-    {
-        uint32_t newOffset;
+    VerifyOrExit((header.GetType() == Header::kTypeResponse) && (header.GetQueryType() == Header::kQueryTypeStandard) &&
+                     !header.IsTruncationFlagSet(),
+                 error = OT_ERROR_DROP);
 
-        SuccessOrExit(error = Name::ParseName(aMessage, offset));
+    aResponse.mQuery = FindQueryById(header.GetMessageId());
+    VerifyOrExit(aResponse.mQuery != nullptr, error = OT_ERROR_NOT_FOUND);
 
-        SuccessOrExit(error = aMessage.Read(offset, record));
+    info.ReadFrom(*aResponse.mQuery);
+    aType = info.mQueryType;
 
-        if (record.Matches(ResourceRecord::kTypeAaaa))
-        {
-            // Return the first found IPv6 address.
-            FinalizeDnsTransaction(*message, queryMetadata, &record.GetAddress(), record.GetTtl(), OT_ERROR_NONE);
-            ExitNow(error = OT_ERROR_NONE);
-        }
+    // Check the Question Section
+
+    VerifyOrExit(header.GetQuestionCount() == kQuestionCount[aType], error = OT_ERROR_PARSE);
```

Comment:
Is this check too strict? We have seen DNS servers that do not echo back any question. 

Besides that, I think a good benefit of not echoing back questions in Thread networks is reducing packet size.

Reply:
I am fine relaxing this check if it is valid/permitted by spec for server to not include question the response.

I tried doing a quick search for this in specs/rfcs, but could not find anything explicit about this. @simonlingoogle, do you happen to know if this is mentioned in any rfc.

I will relax the check in the code for now.

Reply:
RFC 1035 Section 7.3 has the following statement:
```
The next step is to match the response to a current resolver request.
The recommended strategy is to do a preliminary matching using the ID
field in the domain header, and then to verify that the question section
corresponds to the information currently desired.
```
This would imply to me that the resolver expects the question section to include the question that the name server is answering. However, I admit that I'm not sure why the question section is necessary in a response, given that there is an ID field.

@simonlingoogle , can you reference existing implementations that do not include the question in the response? If there is sufficient data that deployed implementations exclude the question, then I think it is something we can explore.

Reply:
I think what I saw is mDNS responses which "MUST NOT contain any questions in the Question Section". So, unicast responses should still contain questions.

Reply:
I did actually relax this check in current PR. Let me change that back.